explorer/interpreter/type_checker.cpp

```diff
@@ -4462,6 +4464,93 @@ auto TypeChecker::TypeCheckCallableDeclaration(Nonnull<CallableDeclaration*> f,
   return Success();
 }
 
+auto TypeChecker::CheckPatternEquality(Nonnull<const Pattern*> o_patt,
```

Comment:
Looking a bit more into function, we are using `Interpreter::PatternMatch` to check arguments: Would this help simplify some of the logic here?

It is for example used in `Interpreter::CallFunction` after all conversions to arguments. It might not be sufficient itself, but I think it is likely usable, possibly with some modifications.

Reply:
The way this is [done for `impl`s of `interface`s](https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/type_checker.cpp#L5024) is by comparing the type of the function in the `impl` against the type of the function in the `interface` (after performing any substitutions if the `interface` is parameterized). I think that's (approximately) the rule we want for virtual functions too: it should be OK for the pattern to be different so long as it matches the same type. Eg:

```
base class A(T:! type) {
  virtual fn F(x: T);
}
class B extends A((i32, i32)) {
  impl fn F((a: i32, b: i32));
}
```

... seems reasonable to me.

Comparing the function types would not be quite right here: we want to allow the `self` type to be the corresponding subtype of the `self` type in the base class (eg, `Base` -> `Derived` or `Base*` -> `Derived*`, but not `Base*` -> `OtherDerived*`), so we will need to separately compare the function parameter types and perform some kind of custom comparison for the receiver types (if any).

We might also want to allow the function type to otherwise be a supertype of the base class function's type (that is, accepting a wider set of parameter types and returning a narrower type), but I don't think we have a design direction there yet, so leaving that out initially makes sense to me.